src/key_value_store.rs

```diff
@@ -206,13 +213,18 @@ impl<K: Clone + Eq + Hash, V: Clone> Persist<(K, V)> for PersistableHashMap<K, V
     }
 
     fn insert(&mut self, change: (K, V)) {
-        self.index.insert(change.0.clone(), change.1.clone());
-        self.pending_changes.push(IndexChange::Add(change));
+        if let Some(old) = self.index.insert(change.0.clone(), change.1.clone()) {
+            self.pending_changes
+                .push(IndexChange::Update((change.0, old)))
+        } else {
+            self.pending_changes.push(IndexChange::Add(change));
+        }
     }
 
     fn remove(&mut self, change: (K, V)) {
-        self.index.remove(&change.0);
-        self.pending_changes.push(IndexChange::Remove(change));
+        if self.index.remove(&change.0).is_some() {
+            self.pending_changes.push(IndexChange::Remove(change));
```

Comment:
Should be `Remove((change.0, old))`, I think. So that when we revert it, we are sure to add back in the old value.

Reply:
I guess I made the assumption that we were always passing in the actual value being removed but no reason to assume that is the case.

Reply:
If that's true then fine, but we should probably assert that or something.

Reply:
Maybe we should use the key rather than the `change` tuple as the input, and always get the old value from the index.

Reply:
Oh but this will complicate the type of `change`, since for `PersistableHashSet` it's just `K` but for other structures it's a tuple. I guess it's good as is.